### PR TITLE
Upgrade Pinot Connector Libraries to 0.8.0

### PR DESCRIPTION
**Pull request by elonazoulay**

### CODE REVIEW

**Comment by elonazoulay** [comment]
Based on #7630 .

**Comment by hashhar** [comment]
Does the upgrade require the changes from https://github.com/trinodb/trino/pull/7630?
If not, let's extract it so that we can merge this quickly.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 1]
```diff
@@ -64,53 +68,62 @@ public static DynamicTable buildFromPql(PinotMetadata pinotMetadata, SchemaTable
...
         ImmutableList.Builder<PinotColumnHandle> aggregateColumnsBuilder = ImmutableList.builder();
-        if (request.getAggregationsInfo() != null) {
+        if (queryContext.getAggregationFunctions() != null) {
+            checkState(queryContext.getAggregationFunctions().length > 0, "Aggregation Functions is empty");
             for (AggregationFunction aggregationFunction : queryContext.getAggregationFunctions()) {
                 aggregateColumnsBuilder.add(new PinotColumnHandle(
                         aggregationFunction.getResultColumnName(),
                         toTrinoType(aggregationFunction.getFinalResultColumnType())));
             }
+            if (queryContext.getGroupByExpressions() != null) {
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:107]: Why is this nested inside the aggregations block?
I think it's possible for queries to have GROUP BY without aggregations.

`SELECT 1 FROM t GROUP BY c`?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 2]
```diff
@@ -64,53 +68,62 @@ public static DynamicTable buildFromPql(PinotMetadata pinotMetadata, SchemaTable
...
+                groupByColumns = getPinotColumnNames(schemaTableName, queryContext.getGroupByExpressions(), columnHandles);
+            }
+        }
+        else {
+            checkState(queryContext.getSelectExpressions() != null && !queryContext.getSelectExpressions().isEmpty(), "Pinot selections is null or empty");
+            checkState(queryContext.getAliasList() == null || queryContext.getAliasList().isEmpty() || !queryContext.getAliasList().stream().anyMatch(alias -> alias != null), "Aliases are not supported");
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:113]: Seems like we couldn't check this earlier but do now. Correct?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 3]
```diff
@@ -142,41 +155,55 @@ private static Type toTrinoType(DataSchema.ColumnDataType columnDataType)
...
+                return ExpressionContext.forFunction(new FunctionContext(expressionContext.getFunction().getType(), expressionContext.getFunction().getFunctionName(), arguments));
+            default:
+                throw new PinotException(PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "Unsupported pinot expression type: " + expressionContext.getType());
         }
     }
 
-    private static OptionalLong getOffset(BrokerRequest request)
+    private static OptionalLong getOffset(QueryContext queryContext)
     {
-        if (request.getSelections() != null && request.getSelections().getOffset() > 0) {
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:190]: Instead of removing the check let's add a `checkState/verify` here since there may be other callers of this method.

Or does the Pinot library populate offset correctly even when no selections are available?

**Review by hashhar — COMMENTED** [review]
First pass. Didn't review PinotSqlFormatter yet.

Would appreciate someone familiar with Pinot to also review the changes in DynamicTableBuilder (and PinotSqlFormatter).

File: plugin/trino-pinot/pom.xml [hunk 4]
```diff
@@ -382,6 +384,13 @@
             <scope>runtime</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.pinot</groupId>
+            <artifactId>pinot-yammer</artifactId>
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/pom.xml:389]: This is used by the new PinotMetricsFactory class which is used by PinotQueryClient. This is a new dependency due to pinot 0.8.0 upgrade.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java [hunk 5]
```diff
@@ -303,8 +305,11 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
             Map<ColumnHandle, Domain> supported = new HashMap<>();
             Map<ColumnHandle, Domain> unsupported = new HashMap<>();
             for (Map.Entry<ColumnHandle, Domain> entry : domains.entrySet()) {
+                Type type = ((PinotColumnHandle) entry.getKey()).getDataType();
+                // Do not push down filters for arrays or varbinary columns:
                 // Pinot does not support array literals
-                if (((PinotColumnHandle) entry.getKey()).getDataType() instanceof ArrayType) {
+                // Pinot does not fully support varbinary null literals (empty byte array): https://github.com/apache/pinot/issues/4230
+                if (type instanceof ArrayType || type instanceof VarbinaryType) {
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotMetadata.java:312]: I added a fix and reverted this change. Also added a test to verify: `TestPinotIntegrationSmokeTest::testVarbinaryFilters`

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 6]
```diff
@@ -64,53 +68,62 @@ public static DynamicTable buildFromPql(PinotMetadata pinotMetadata, SchemaTable
...
+            orderBy = orderByBuilder.build();
         }
 
-        List<String> groupByColumns;
-        if (request.getGroupBy() == null) {
-            groupByColumns = ImmutableList.of();
-        }
-        else {
-            groupByColumns = resolvePinotColumns(schemaTableName, request.getGroupBy().getExpressions(), columnHandles);
+        Optional<String> filter = Optional.empty();
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:90]: I reverted the order of processing so the functionality is the same as it was with the following caveat:
A passthrough query with grouping columns but no aggregations is parsed by pinot 0.8.0 differently than other queries: it wraps the grouping columns in a `distinct(col1, col2, ...)` select list, returns a null list of grouping columns and a "distinct" aggregation function which does not support the `Pinot AggregatationFunction::getFinalResultColumnType` method. 

I addressed this but it seems better to include it in a separate pr, as the previous functionality is preserved with this pr and rewriting would change how distinct aggregations are handled. wdyt?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 1 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:107]: reverted.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 2 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:113]: Yes, the new QueryContext has an alias list which we did not use before (pinot is deprecating/eliminating broker request "pql" and moving it all to QueryContext). I've also added support for aliases but will put it in a subsequent pr as it changes functionality.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 3 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:190]: In pinot-0.8.0 all the request.get# methods return null. I moved them to use the QueryContext api.

File: plugin/trino-pinot/pom.xml [hunk 7]
```diff
@@ -14,7 +14,7 @@
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/pom.xml:14]: Removed. Also added context in the conversation, i.e. previously pinot created string arrays when a column was defined as a boolean array but now it fails. So technically it was never supported.

**Comment by elonazoulay** [comment]
Fixes #9133

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 1 above)
Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java]: Let's add a test with this query:

1. Ensure that the plan contains a TableScan (so that we are sure that the connector participated in the query and the engine didn't do it all itself).
2. The results are correct.

Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java]: I created https://github.com/trinodb/trino/issues/9167 to test this in BaseConnectorTest as well.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/DoubleDecoder.java [hunk 8]
```diff
@@ -29,6 +29,10 @@ public void decode(Supplier<Object> getter, BlockBuilder output)
         if (value == null) {
             output.appendNull();
         }
+        else if (value instanceof String) {
+            // Pinot returns NEGATIVE_INFINITY, POSITIVE_INFINITY as a String
+            DOUBLE.writeDouble(output, Double.valueOf(value.toString()));
+        }
         else {
             DOUBLE.writeDouble(output, ((Number) value).doubleValue());
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/DoubleDecoder.java:37]: Seeing a cast to `Number` makes me sad - can we make this stricter?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java [hunk 9]
```diff
@@ -157,6 +158,9 @@ private static Object convertValue(Type type, Object value)
         else if (type instanceof VarbinaryType) {
             return Hex.encodeHexString(((Slice) value).getBytes());
         }
+        else if (type instanceof VarcharType) {
+            return ((Slice) value).toStringUtf8();
+        }
         return value;
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:164]: We are missing quite a lot of types here it seems?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java [hunk 10]
```diff
@@ -95,7 +95,7 @@ protected QueryRunner createQueryRunner()
                     createTestRecord(
                             Arrays.asList("string_" + (offset), "string1_" + (offset + 1), "string2_" + (offset + 2)),
                             Arrays.asList(false, true, true),
-                            Arrays.asList(54, -10001, 1000),
+                            Arrays.asList(54 + i / 3, -10001, 1000),
```
Comment (hashhar) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java:98]: Is this to get a fractional value?

Comment (hashhar) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java:98]: Also I'm confused by the fact that this column has integers, booleans, string and fractional values.

What should I look at to figure out how they actually land into Pinot? The table config?

**Review by hashhar — COMMENTED** [review]
Let's extract until https://github.com/trinodb/trino/pull/9098/commits/5fc2a82e1c9f39f13a1e64b4fdbb6be3d3c42ca9 into a separate PR since those commits are all good and can be merged right now.

Let's also extract the below three into another PR since those look good and can be merged:
Add support for varbinary filters 93a80ca
Fix decoding real -INFINITY and +INFINITY e88b097
Fix decoding double -INFINITY and +INFINITY 82f2914

Let's extract "Fix in clause handling in filter pushdown e6ff226" into a separate PR (for the sake of release notes and people who'd want to look at the fix).

We can then focus on just the 0.8.0 upgrade commits here (which I still need to review some parts of).

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/DoubleDecoder.java (hunk 8 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/DoubleDecoder.java:37]: This code was unchanged from before, the reason: pinot transform functions declare that they return double (for example) but the actual value can be a double, long, int, or float. Let me know if you have any suggestions.

Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/DoubleDecoder.java:37]: The problem with Number is that it applies to things like BigDecimal too.
We can address this separately as part of larger effort to remove Number casts from the code where possible.

Let's keep as is for now.

**Comment by elonazoulay** [comment]
Created https://github.com/trinodb/trino/pull/9178 for the first pr

**Comment by elonazoulay** [comment]
Created https://github.com/trinodb/trino/pull/9180 for the varbinary filter support + real/double fixes

**Comment by elonazoulay** [comment]
Pr for the in clause varchar handling: https://github.com/trinodb/trino/pull/9181

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java (hunk 10 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/TestPinotIntegrationSmokeTest.java:98]: This is just to have more distinct values in the column instead of just `54`, it is an integer so the values would be `54`, `54`, `54`, `55`, `55`, `55`, etc.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java (hunk 9 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:164]: The rest of the supported types were already working, there are tests. This is just to add support for varbinary. For context the other supported types are primitives (arrays are not pushed down as pinot does not have a way to express an array literal): bigint, double, integer, real, boolean. Let me know if you have any suggestions.

Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotQueryBuilder.java:164]: Continued at https://github.com/trinodb/trino/pull/9181#discussion_r705497387

File: plugin/trino-pinot/pom.xml [hunk 11]
```diff
@@ -382,6 +384,13 @@
             <scope>runtime</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.pinot</groupId>
+            <artifactId>pinot-yammer</artifactId>
+            <version>${dep.pinot.version}</version>
+            <scope>runtime</scope>
```
[the remark this answers is not in the document]
Comment (xiangfu0) [reply plugin/trino-pinot/pom.xml:403]: shall we make a noop implementation for this pinot-metrics interface ?

[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/pom.xml:403]: > how are we getting this during runtime?

Required by the PinotQueryClient which requires PinotMetricsFactory. Would it be ok to leave it, so that we can have broker metrics? The worker functions like a broker without a reducer service (soon we will put that in also). wdyt?

**Review by hashhar — COMMENTED** [review]
Some more comments and questions

Still need to review more things.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java [hunk 12]
```diff
@@ -0,0 +1,687 @@
...
+    static {
+        Map<Predicate.Type, Rule<Predicate>> predicateMap = new HashMap<>();
+        predicateMap.put(Predicate.Type.IN, new ValuesListPredicateRule(Predicate.Type.IN, "IN"));
+        predicateMap.put(Predicate.Type.NOT_IN, new ValuesListPredicateRule(Predicate.Type.NOT_IN, "NOT IN"));
+        predicateMap.put(Predicate.Type.RANGE, new RangePredicateRule());
+        predicateMap.put(Predicate.Type.REGEXP_LIKE, new BinaryFunctionPredicateRule(Predicate.Type.REGEXP_LIKE, "regexp_like"));
+        predicateMap.put(Predicate.Type.TEXT_MATCH, new BinaryFunctionPredicateRule(Predicate.Type.TEXT_MATCH, "text_match"));
+        predicateMap.put(Predicate.Type.JSON_MATCH, new BinaryFunctionPredicateRule(Predicate.Type.JSON_MATCH, "json_match"));
+        predicateMap.put(Predicate.Type.IS_NULL, new ExpressionOnlyPredicate(Predicate.Type.IS_NULL, "IS NULL"));
+        predicateMap.put(Predicate.Type.IS_NOT_NULL, new ExpressionOnlyPredicate(Predicate.Type.IS_NOT_NULL, "IS NOT NULL"));
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java:109]: Definitely. Adding tests for these.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java [hunk 13]
```diff
@@ -0,0 +1,687 @@
...
+        PREDICATE_RULE_MAP = immutableEnumMap(predicateMap);
+
+        Map<TransformFunctionType, Rule<FunctionContext>> functionMap = new HashMap<>();
+        functionMap.put(CASE, new CaseFunctionRule());
+        functionMap.put(CAST, new CastFunctionRule());
+        functionMap.put(MINUS, new MinusFunctionRule());
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java:115]: It's the way pinot currently parses. There are 3 types of expressions: function, identifier, literal.
If pinot sees `function|identifier` <binary operator> `function|identifier` it changes it to `arg1 - arg2 <operator> 0`.
This leads to correctness problems so this has to be rewritten back to the original form.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java [hunk 14]
```diff
@@ -0,0 +1,687 @@
...
+        public String formatToSql(Predicate object, Captures captures, Context context)
+        {
+            RangePredicate rangePredicate = (RangePredicate) object;
+            ExpressionContext predicateExpression = captures.get(PREDICATE_EXPRESSION);
+            String expression = formatExpression(predicateExpression, context);
+
+            // Single value range should have been rewritten in formatBinaryOperatorPredicate
+            checkState(!rangePredicate.getLowerBound().equals(UNBOUNDED) && !rangePredicate.getUpperBound().equals(UNBOUNDED), "Unexpected range predicate '%s'", rangePredicate);
+            if (rangePredicate.isUpperInclusive() && rangePredicate.isLowerInclusive()) {
+                return format("(%s) BETWEEN %s AND %s", expression, singleQuoteValue(rangePredicate.getLowerBound()), singleQuoteValue(rangePredicate.getUpperBound()));
```
[the remark this answers is not in the document]
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/PinotSqlFormatter.java:471]: The pinot parser will rewrite x <=  y and x >= z as between no matter what the input sql is (i.e. even if rewritten to use <= and >=). Pinot parses `x` (>|<|<=|>=) `y` into a RangePredicate class along with between so all the cases need to be considered. Does that make sense?

**Comment by ebyhr** [comment]
@elonazoulay Please rebase on upstream to resolve conflicts. Could you add a new test case for column names having double-quotes https://github.com/trinodb/trino/pull/9459#issuecomment-933121939? I assume the issue is resolved by this upgrade.

**Comment by ddcprg** [comment]
Any chances this gets released with 363?

**Comment by elonazoulay** [comment]
Thanks @ebyhr - due to aliases and group by requiring quotes I addressed this in the commit: "Broker request should submit json", similar idea to your fix.

**Comment by elonazoulay** [comment]
Added bug fixes and cleanups to simplify logic and make the code easier to reason about:
* PinotColumn and PinotExpression are eliminated.
* Logic is simplified for getting column metadata.
* Bugs related to column quoting and aggregation pushdown fixed.

I feel like the commits should be reorganized, as it will make it easier to review (a lot of earlier commits are no longer used).
Unless the entire pr will be squashed I would be happy to do this, let me know: @ebyhr , @hashhar, @findepi.

And a last question: I am removing DynamicTable and PinotQueryBuilder - will simplify logic as the query will only be parsed in one place. Should I do that as a follow up pr? Assuming yes for now (will push it based on this one).

**Review by hashhar — COMMENTED** [review]
For "Remove PinotColumn":

Maybe reword message to `Remove PinotColumn and moved used methods to PinotColumnHandle instead.` to clarify it was a mechanical refactor - nothing actually changed.

File: plugin/trino-pinot/pom.xml [hunk 15]
```diff
@@ -14,7 +14,7 @@
 
     <properties>
         <air.main.basedir>${project.parent.basedir}</air.main.basedir>
-        <dep.pinot.version>0.6.0</dep.pinot.version>
+        <dep.pinot.version>0.8.0</dep.pinot.version>
```
[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-pinot/pom.xml:17]: We decided to drop backward compat according to discussion on Slack. https://trinodb.slack.com/archives/CP1MUNEUX/p1634735523418100

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/AggregateExpression.java [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+
+    @JsonProperty
+    public boolean isReturnNullOnEmptyGroup()
+    {
+        return returnNullOnEmptyGroup;
+    }
+
+    public String toFieldName()
+    {
+        return format("%s(%s)", function, argument);
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/AggregateExpression.java:73]: Do field names not require quoting?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java [hunk 17]
```diff
@@ -72,6 +74,8 @@ public DynamicTable(
         this.limit = requireNonNull(limit, "limit is null");
         this.offset = requireNonNull(offset, "offset is null");
         this.query = requireNonNull(query, "query is null");
+        this.isAggregateInSelectList = selections.stream()
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:77]: nit: `projections` instead of `selections`?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java [hunk 18]
```diff
@@ -134,6 +138,11 @@ public String getQuery()
         return query;
     }
 
+    public boolean isAggregateInSelectList()
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:141]: Do I understand correctly that this tracks if the aggregation column is already part of the projected columns or not?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java [hunk 19]
```diff
@@ -142,41 +150,61 @@ private static Type toTrinoType(DataSchema.ColumnDataType columnDataType)
...
     {
-        ImmutableList.Builder<String> pinotColumnNamesBuilder = ImmutableList.builder();
-        for (String trinoColumnName : trinoColumnNames) {
-            if (trinoColumnName.equals(WILDCARD)) {
-                pinotColumnNamesBuilder.addAll(columnHandles.values().stream().map(handle -> ((PinotColumnHandle) handle).getColumnName()).collect(toImmutableList()));
+        ImmutableList.Builder<PinotColumnHandle> pinotColumnsBuilder = ImmutableList.builder();
+        for (int index = 0; index < expressions.size(); index++) {
+            ExpressionContext expressionContext = expressions.get(index);
+            Optional<String> alias = getAlias(aliases, index);
+            // Only substitute * with columns for top level SELECT *
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:159]: Can you explain why?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java [hunk 20]
```diff
@@ -106,7 +104,7 @@ private static String convertOrderByExpressionToPql(OrderByExpression orderByExp
     {
         requireNonNull(orderByExpression, "orderByExpression is null");
         StringBuilder builder = new StringBuilder()
-                .append(quoteIdentifier(orderByExpression.getColumn()));
+                .append(orderByExpression.getExpression());
```
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:107]: Is this quoted somewhere?

**Review by hashhar — COMMENTED** [review]
For "Upgrade libraries to pinot-0.8.0"

**Review by hashhar — COMMENTED** [review]
Move "Update supported pinot version in documentation" into the commit that updates the library version.

**Comment by hashhar** [comment]
LGTM % comments.

Sorry about the long back and forth for this PR.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java (hunk 18 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:141]: Yep, that's correct.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 19 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:159]: The only times "*" occurs in pql is in `select *` or `count(*)`. Since `select *` is the only time that "*" needs to be substituted with all columns.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/AggregateExpression.java (hunk 16 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/AggregateExpression.java:73]: The expression field is quoted, that's what is used when generating sql. The fieldName becomes the columnName in PinotColumnHandle which is not quoted.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java (hunk 20 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTablePqlExtractor.java:107]: Yes, the OrderByExpression is created from the PinotColumnHandle.getExpression which is quoted.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java (hunk 17 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:77]: sgtm!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java (hunk 18 above)
Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTable.java:141]: Also changed this to `isAggregateInProjections` to be consistent. wdyt?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java (hunk 19 above)
Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:159]: The top-level makes it confusing for me. Does it mean that in a query like `SELECT * FROM (SELECT * FROM tbl)` you'd only substitute the outer `*`? Or is the assumption that the query would fail anyway since Pinot doesn't support subqueries so DynamicTableBuilder would never see a subquery passed to it?

Comment (elonazoulay) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:159]: Yep, pinot does not have subqueries so that query would fail during parsing. Should I remove the "top level" part?

Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java:159]: Instead add something like

`Only substitute * with columns for top level SELECT *. Since Pinot doesn't support subqueries (yet) we can only have one occurence of SELECT *`.

"top-level" was misleading since the code doesn't seem to enforce that.

**Review by hashhar — APPROVED** [review]
Thanks @elonazoulay

**Comment by hashhar** [comment]
CI hit https://github.com/trinodb/trino/issues/8792 and https://github.com/trinodb/trino/issues/8911.
